### PR TITLE
CORTX-29592: Support deploy into non-default namespace

### PR DESCRIPTION
**Pull request by walterlopatka**
## Description
This change allows CORTX to be deployed into any namespace (including default).
* If the namespace does not exist it will be created.
* In the solution.example.yaml file the default namespace is now `cortx`

Note: This does not support multiple CORTX deployments into a single k8s cluster.  That will come later.

Note: I do not consider this to be a breaking change.  The default namespace in `solution.example.yaml` has changed from `default` to `cortx`.  But given an existing `solution.example.yaml` file, which already has a specific namespace specified.  This could possible change behavior of tools that automatically copy `solution.example.yaml` to generate the config file.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds new functionality)
- [ ] Breaking change (bug fix or new feature that breaks existing functionality)
- [ ] Third-party dependency update
- [ ] Documentation additions or improvements
- [ ] Code quality improvements to existing code or test additions/updates

## Applicable issues

- This change is the first to address CORTX-29592

## How was this tested?

* Testing is in progress.

## Checklist

- [x] The change is tested and works locally.
- [x] New or changed settings in the solution YAML are documented clearly in the README.md file.
- [x] All commits are signed off and are in agreement with the [CORTX Community DCO and CLA policy](https://github.com/Seagate/cortx/blob/main/doc/dco_cla.md).

If this change addresses a CORTX Jira issue:

- [x] The title of the PR starts with the issue ID (e.g. `CORTX-XXXXX:`)

### CODE REVIEW

File: k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-configmap/values.yaml
Comment (walterlopatka) [inline k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-configmap/values.yaml:56]: Perhaps these should be shortened to `tcp://consul-server:8301` (and so on).  As currently implemented, these are set by  `deploy-cortx-cloud.sh` as command line args to `helm install`.

File: k8_cortx_cloud/cortx-cloud-3rd-party-pkg/templates/local-path-storage-template.yaml
Comment (walterlopatka) [inline k8_cortx_cloud/cortx-cloud-3rd-party-pkg/templates/local-path-storage-template.yaml:156]: Hmmmm... I have no idea why I would have removed this.  I'll restore it.

File: k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-platform/templates/cortx-hax-svc.yaml
Comment (walterlopatka) [inline k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-platform/templates/cortx-hax-svc.yaml:1]: `cortx-platform` is installed twice as part of the current `deploy-cortx-cloud.sh` script.  This was not an issue previously because "if namespace != default" guarded the first call to create the namespace.

File: k8_cortx_cloud/deploy-cortx-cloud.sh
Comment (walterlopatka) [inline k8_cortx_cloud/deploy-cortx-cloud.sh:1317]: Whoops... I should have removed this.

File: k8_cortx_cloud/destroy-cortx-cloud.sh
Comment (walterlopatka) [inline k8_cortx_cloud/destroy-cortx-cloud.sh:394]: I commented out this entire function.  It's not obvious to me that it is needed, though I am not 100% sure yet.  I need to review more closely and make a final determination.

File: k8_cortx_cloud/destroy-cortx-cloud.sh
Comment (walterlopatka) [inline k8_cortx_cloud/destroy-cortx-cloud.sh:435]: I'm not sure about this... I need to review this and consider it more carefully.

File: k8_cortx_cloud/destroy-cortx-cloud.sh
Comment (walterlopatka) [inline k8_cortx_cloud/destroy-cortx-cloud.sh:490]: Need to delete these commented out "if / fi" lines.

File: k8_cortx_cloud/deploy-cortx-cloud.sh
Comment (walterlopatka) [inline k8_cortx_cloud/deploy-cortx-cloud.sh:3]: Whoops... need to remove this.

File: k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-platform/templates/cortx-io-svc.yaml
Comment (keithpine) [inline k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-platform/templates/cortx-io-svc.yaml:3]: Could you make this fix too? (see #206)

```suggestion
{{- if gt (int $svc.count) 0 -}}
```

Comment (walterlopatka) [reply k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-platform/templates/cortx-io-svc.yaml:3]: Yes, certainly!  I'll make sure that's in the net update.

File: k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-platform/templates/cortx-hax-svc.yaml
Comment (walterlopatka) [reply k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-platform/templates/cortx-hax-svc.yaml:1]: Since posting this comment, there has been a change in that cortx-platform is now only helm-installed once, but this code remains until/unless we decide to remove the "create" field throughout this helm chart.

File: k8_cortx_cloud/deploy-cortx-cloud.sh
Comment (walterlopatka) [reply k8_cortx_cloud/deploy-cortx-cloud.sh]: done

File: k8_cortx_cloud/destroy-cortx-cloud.sh
Comment (walterlopatka) [reply k8_cortx_cloud/destroy-cortx-cloud.sh]: This has since been removed.

File: k8_cortx_cloud/destroy-cortx-cloud.sh
Comment (walterlopatka) [reply k8_cortx_cloud/destroy-cortx-cloud.sh]: I restored this one.

File: k8_cortx_cloud/destroy-cortx-cloud.sh
Comment (walterlopatka) [reply k8_cortx_cloud/destroy-cortx-cloud.sh]: done

File: k8_cortx_cloud/deploy-cortx-cloud.sh
Comment (walterlopatka) [reply k8_cortx_cloud/deploy-cortx-cloud.sh]: removed

File: k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-platform/templates/cortx-io-svc.yaml
Comment (walterlopatka) [reply k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-platform/templates/cortx-io-svc.yaml]: This has been addressed.

**Review by osowski — COMMENTED** [review]
Overall, these changes look great. Two comments that we can discuss more next week.

1. We can change the default create/delete practice to install Rancher with the `kubectl apply` command, instead of `kubectl create`, while also not deleting Rancher on a standard destroy script run. We can add a flag to the destroy script to force deletion of Rancher, but I don't think it's that much of an issue to leave Rancher deployed, as long as we use the `kubectl apply` method to create it.
2. We can remove the `namespace` element in each of the Helm Charts and instead of using `.Values.namespace`, we can use `.Release.Namespace` as it is always provided by Helm and we are explicitly passing in the `--namespace` to the `helm install` commands. I have already done this for CORTX-28968. We don't need to include `metadata.namespace` as a hard-coded namespace for objects, since the previous application of `.Release.Namespace` already does that for us. The less code we can write to do the same thing, the better.

File: README.md
Comment (keithpine) [inline README.md:209]: Forgive me for commenting on something that's already existing, but two nitpicks:
* We deploy third party resources (Consul, etc.) into the same namespace, not just CORTX.
* It's also not just Pods.

File: k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-platform/templates/cortx-hax-svc.yaml
Comment (keithpine) [reply k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-platform/templates/cortx-hax-svc.yaml:1]: I think it's fine, we can fix it later.

**Review by keithpine — COMMENTED** [review]
Looking good. I agree with Rick's comments.

**Comment by keithpine** [comment]
Thought I'd mention, I had a manual install of consul in a non-default namespace. When I deployed CORTX into another namespace, the consul install failed. Not sure if it's related to my other namespace install or not.

**Comment by walterlopatka** [comment]
I have addressed the comments I have received.  Specifically:
* I have removed the "namespace" values in the cortx helm charts, instead relying on .Release.Namespace to provide that information.
* I have changed "kubectl create" local-storage to "kubectl apply" local-storage, and removed the corresponding "delete" local storage.  (I also added a comment to the "undeploy" section of the doc noting how one can delete the local storage.  I thought this was better than adding a bash option.)
* I addressed Keith's legit comment about the namespace documentation

I have tested these changes, but I will do more testing early next week before merging.

(I have not yet looked into Keith's comment about deploying with consul already existing in another namespace.  I'll look at that on Monday.)

File: README.md
Comment (keithpine) [inline README.md:218]:
```suggestion
| `namespace`       | The Kubernetes namespace that all CORTX-related resources will be deployed into.
```

File: k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-platform/templates/cortx-io-svc.yaml
Comment (keithpine) [inline k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-platform/templates/cortx-io-svc.yaml:13]:
```suggestion
  namespace: {{ $root.Release.Namespace }}
```

File: k8_cortx_cloud/deploy-cortx-cloud.sh
Comment (keithpine) [inline k8_cortx_cloud/deploy-cortx-cloud.sh:403]:
```suggestion
                                               --namespace "${namespace}"
```

There's not a consistent use of `-n` and `--namespace` in this script already. My preference is to always use long options in scripts since they are self documenting.

File: k8_cortx_cloud/deploy-cortx-cloud.sh
Comment (keithpine) [inline k8_cortx_cloud/deploy-cortx-cloud.sh:405]:
```suggestion
                                               --namespace "${namespace}"
```

File: k8_cortx_cloud/deploy-cortx-cloud.sh
Comment (keithpine) [inline k8_cortx_cloud/deploy-cortx-cloud.sh:409]:
```suggestion
    kubectl rollout restart statefulset/consul-server --namespace "${namespace}"
    kubectl rollout restart daemonset/consul-client --namespace "${namespace}"
```

File: k8_cortx_cloud/deploy-cortx-cloud.sh
Comment (keithpine) [inline k8_cortx_cloud/deploy-cortx-cloud.sh:913]:
```suggestion
    if ! kubectl wait --for=condition=available --timeout="${TIMEOUT}" --namespace "${NAMESPACE}" "$@"; then
```

File: k8_cortx_cloud/deploy-cortx-cloud.sh
Comment (keithpine) [inline k8_cortx_cloud/deploy-cortx-cloud.sh:915]:
```suggestion
        if ! kubectl wait --for=condition=available --timeout="${TIMEOUT}" --namespace "${NAMESPACE}" "$@"; then
```

File: k8_cortx_cloud/deploy-cortx-cloud.sh
Comment (keithpine) [inline k8_cortx_cloud/deploy-cortx-cloud.sh:1337]:
```suggestion
       kubectl get secrets/${cortx_secret_name} --namespace ${namespace} --template={{.data.s3_auth_admin_secret}} | base64 -d
```

File: k8_cortx_cloud/deploy-cortx-cloud.sh
Comment (keithpine) [inline k8_cortx_cloud/deploy-cortx-cloud.sh:1342]:
```suggestion
       kubectl get secrets/${cortx_secret_name} --namespace ${namespace} --template={{.data.csm_mgmt_admin_secret}} | base64 -d"
```

File: k8_cortx_cloud/destroy-cortx-cloud.sh
Comment (keithpine) [inline k8_cortx_cloud/destroy-cortx-cloud.sh:323]:
```suggestion
        done < <(kubectl get pods --namespace "${namespace}" | grep -e kafka -e zookeeper -e openldap -e consul 2>&1)
```

File: k8_cortx_cloud/destroy-cortx-cloud.sh
Comment (keithpine) [inline k8_cortx_cloud/destroy-cortx-cloud.sh:342]:
```suggestion
            kubectl patch pvc --namespace "${namespace}" "${volume_claim}" -p '{"metadata":{"finalizers":null}}'
```

File: k8_cortx_cloud/destroy-cortx-cloud.sh
Comment (keithpine) [inline k8_cortx_cloud/destroy-cortx-cloud.sh:344]:
```suggestion
        kubectl delete pvc --namespace "${namespace}" "${volume_claim}"
```

**Review by keithpine — APPROVED** [review]
Just some nitpick comments, fix them if you want, otherwise, LGTM.

**Comment by walterlopatka** [comment]
Thanks @keithpine for the thorough review.  I have accepted all those suggestions.  I think it's ready to go.  I'll give @osowski a shot at reviewing this later today if he wants.

**Review by keithpine — APPROVED** [review]

**Review by osowski — APPROVED** [review]
lgtm!

**Pull request merged**
